### PR TITLE
Fix naming scheme

### PR DESCRIPTION
**Pull request by makeclean**
This PR introduces the DAGMC Metadata Class providing a unified location where we can store all the metadata for a given DAGMC problem. This allows every MC code to use the class, providing a common method for metadata access.

### CODE REVIEW

File: mcnp/mcnp_funcs.cpp [hunk 1]
```diff
@@ -283,317 +141,144 @@ void dagmcwritemcnp_(char* dagfile, char *lfile, int *llen)  // file with cell/s
...
-      std::cout << cellid << " has the following density assignments" << std::endl;
-      for ( int j = 0 ; j < density_props.size() ; j++ ) {
-        std::cout << density_props[j] << std::endl;
+    } else {
+      std::string mat_name = DMD->volume_material_property_data_eh[entity];
+      // if we not vacuum or graveyard
+      if(mat_name.find("Vacuum") == std::string::npos || mat_name.find("Graveyard") == std::string::npos) {
+        pyne::Material material = workflow_data.material_library[mat_name];
+        mat_num = material.metadata["mat_number"].asString();
+        density = "-1.0"+_to_string(material.density);
```
Comment (ljacobson64) [inline mcnp/mcnp_funcs.cpp]: Should be just "-" instead of "-1.0"

File: common/dagmcmetadata.cpp [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+  keyword_synonyms[ "rho" ] = "density";
+  keyword_synonyms[ "mat" ] = "material";
+}
+
+// destructor doesnt need to do anything
+dagmcMetaData::~dagmcMetaData()
+{
+}
+
+// load the property data from the dagma instance
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: dagmc

File: common/dagmcmetadata.cpp [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+// parse the importance data from the file
+void dagmcMetaData::parse_importance_data()
+{
+  std::map<moab::EntityHandle,std::vector<std::string> > importance_assignments;
+  importance_assignments = get_property_assignments("importance",3,":");
+
+  int num_vols = DAG->num_entities( 3 );
+
+  std::vector<std::string> importance_assignment; // boundary conditions for the current entity
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: importances not boundary conditions

File: common/dagmcmetadata.cpp [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+
+// parse the importance data from the file
+void dagmcMetaData::parse_tally_volume_data()
+{
+  std::map<moab::EntityHandle,std::vector<std::string> > tally_assignments;
+  tally_assignments = get_property_assignments("tally",3,":");
+
+  int num_vols = DAG->num_entities( 3 );
+
+  std::vector<std::string> tally_assignment; // boundary conditions for the current entity
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: volume tallies not boundary conditions

File: common/dagmcmetadata.cpp [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+          std::cout << particle << " assuming importance 1.0 " << std::endl;
+        }
+        // give this particle default importance
+        importance_map[eh][particle] = 1.0;
+      }
+    }
+  }
+}
+
+// parse the importance data from the file
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: volume tallies not importances

File: common/dagmcmetadata.cpp [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+    moab::EntityHandle eh = DAG->entity_by_index( 2, i );
+
+    boundary_assignment = boundary_assignments[eh];
+    if (boundary_assignment.size() != 1 ) {
+      std::cout << "More than one boundary conditions specified for " << surfid <<std::endl;
+      std::cout << surfid << " has the following density assignments" << std::endl;
+      for ( int j = 0 ; j < boundary_assignment.size() ; j++ ) {
+        std::cout << boundary_assignment[j] << std::endl;
+      }
+      std::cout << "Please check your boundary condition assignments " << surfid << std::endl;
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp:338]: should this fail if it finds 2 boundary conditions?

File: common/dagmcmetadata.cpp [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+    // ie. both surfaces and its members triangles,
+
+    if(boundary_assignment[0].find("Reflecting") != std::string::npos )
+      surface_boundary_data_eh[eh] = "Reflecting";
+    if (boundary_assignment[0].find("White") != std::string::npos )
+      surface_boundary_data_eh[eh] = "White";
+  }
+}
+
+// parse the importance data from the file
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: surface tallies not importances

File: common/dagmcmetadata.cpp [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+void dagmcMetaData::parse_tally_surface_data()
+{
+  std::map<moab::EntityHandle,std::vector<std::string> > tally_assignments;
+  tally_assignments = get_property_assignments("tally",2,":");
+
+  int num_surfaces = DAG->num_entities( 2 );
+
+  std::vector<std::string> tally_assignment; // boundary conditions for the current entity
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: surface tallies not boundary conditions

File: common/dagmcmetadata.cpp [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+
+    // set the value of each string for
+    std::string importances = "|";
+    for ( int j = 0 ; j < importance_assignment.size() ; j++ ) {
+      if(importance_assignment[j] == "" ) break;
+      // delimit each particle/value pair with a pipe symbol
+      importances += importance_assignment[j]+"|";
+      // also split to get key-value
+      std::pair<std::string,std::string> pair = split_string(importance_assignment[j],"/");
+      // add to the unique collection of particle names
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp:256]: what happens if it finds 2 importances for the same particle type?

File: common/dagmcmetadata.cpp [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+  // resize the array
+  new_properties.resize(properties_set.size());
+
+  // turn set back into vector
+  std::copy(properties_set.begin(),properties_set.end(),new_properties.begin());
+
+  return new_properties;
+}
+
+// from a given set remove any matches if they are found to keep the
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: ?

File: common/test/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+# Libraries                                                                                                                             
+set(LINK_LIBS)
+list(APPEND LINK_LIBS gtest)
+list(APPEND LINK_LIBS pthread)
+list(APPEND LINK_LIBS dagmciface)
+list(APPEND LINK_LIBS ${DAG_LINK_LIBS})
+
+# Include directories                                                                                                                   
+include_directories(..)
+include_directories(${CMAKE_SOURCE_DIR}/common)
```
Comment (ljacobson64) [inline common/test/CMakeLists.txt:12]: This is the same directory as ..

File: common/test/dagmciface_unit_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+      EXPECT_EQ(mat_prop,impl_comp_prop);
+
+    mat_prop = dgm->get_volume_property("material",i,true);
+
+    if (!DAG->is_implicit_complement(eh))
+      EXPECT_EQ(mat_prop,base_property);
+    else
+      EXPECT_EQ(mat_prop,impl_comp_prop);
+
+    int cellid = DAG->id_by_index( 3, i );
```
Comment (ljacobson64) [inline common/test/dagmciface_unit_tests.cpp]: This appears to be the same test as the previous one, and "cellid" is unused. Perhaps one of these lines was meant to be changed?

File: common/test/dagmciface_unit_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+  int num_vols = DAG->num_entities(3);
+  for ( int i = 1 ; i <= num_vols ; i++ ) {
+    moab::EntityHandle eh = DAG->entity_by_index(3,i);
+    std::string mat_prop = dgm->get_volume_property("density",eh);
+    EXPECT_EQ(mat_prop,base_property);
+
+    mat_prop = dgm->get_volume_property("density",i,true);
+    EXPECT_EQ(mat_prop,base_property);
+
+    int cellid = DAG->id_by_index( 3, i );
```
Comment (ljacobson64) [inline common/test/dagmciface_unit_tests.cpp]: This appears to be the same test as the previous one, and "cellid" is unused. Perhaps one of these lines was meant to be changed?

File: common/test/dagmciface_unit_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+      EXPECT_EQ(mat_prop,impl_comp_prop);
+
+    mat_prop = dgm->get_volume_property("material_density",i,true);
+
+    if (!DAG->is_implicit_complement(eh))
+      EXPECT_EQ(mat_prop,base_property);
+    else
+      EXPECT_EQ(mat_prop,impl_comp_prop);
+
+    int cellid = DAG->id_by_index( 3, i );
```
Comment (ljacobson64) [inline common/test/dagmciface_unit_tests.cpp]: This appears to be the same test as the previous one, and "cellid" is unused. Perhaps one of these lines was meant to be changed?

File: fluka/build/CMakeLists.txt [hunk 15]
```diff
@@ -8,10 +8,14 @@ set(SRC_FILES mainFluDAG.cpp)
 # Libraries
 set(LINK_LIBS)
 list(APPEND LINK_LIBS fludag)
+message(STATUS ${FLUKA_DIR})
```
Comment (ljacobson64) [inline fluka/build/CMakeLists.txt]: Do we need this status message?

File: geant4/build/CMakeLists.txt [hunk 16]
```diff
@@ -1,14 +1,26 @@
 message(STATUS "Building DAG-Geant4")
 
-# Get Geant4 libraries, include directories, and compile definitions
-include(${Geant4_USE_FILE})
-
 # CXX compiler flags
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${Geant4_CXX_FLAGS}")
```
Comment (ljacobson64) [inline geant4/build/CMakeLists.txt:4]: If we're not including the Geant4_USE_FILE here anymore, then Geant4_CXX_FLAGS won't be defined here. Perhaps move this below the new location of the include?

File: common/CMakeLists.txt [hunk 17]
```diff
@@ -1,10 +1,12 @@
...
 # Source files
-set(DI_SRC_FILES ${DI_SRC_FILES} dagmciface.cpp PARENT_SCOPE)
-
+list(APPEND DI_SRC_FILES dagmcmetadata.cpp)
+                                
 # Public headers
-set(DI_PUB_HEADERS ${DI_PUB_HEADERS} dagmciface.h PARENT_SCOPE)
+list(APPEND DI_PUB_HEADERS dagmcmetadata.hpp)
+
+message(STATUS ${DI_SRC_FILES})
```
Comment (ljacobson64) [inline common/CMakeLists.txt]: Do we need this message?

File: common/CMakeLists.txt [hunk 18]
```diff
@@ -1,10 +1,12 @@
 message(STATUS "Building dagmciface library")
 
 # Source files
-set(DI_SRC_FILES ${DI_SRC_FILES} dagmciface.cpp PARENT_SCOPE)
-
+list(APPEND DI_SRC_FILES dagmcmetadata.cpp)
+                                
 # Public headers
-set(DI_PUB_HEADERS ${DI_PUB_HEADERS} dagmciface.h PARENT_SCOPE)
+list(APPEND DI_PUB_HEADERS dagmcmetadata.hpp)
```
Comment (ljacobson64) [inline common/CMakeLists.txt:7]: If dagmciface.cpp and dagmciface.h aren't being included in libdagmciface anymore, should they still exist?

Also why the change from a string to a list? DI_SRC_FILES and DI_PUB_HEADERS are still strings everywhere else they're used

File: mcnp/CMakeLists.txt [hunk 19]
```diff
@@ -1,5 +1,7 @@
 message(STATUS "Including MCNP in dagmciface library")
 
+include_directories(${CMAKE_SOURCE_DIR}/common)
```
Comment (ljacobson64) [inline mcnp/CMakeLists.txt]: I don't think this directory needs to be included here because this CMakeLists.txt isn't actually building anything

File: mcnp/test/dagmcnp_unit_tests.cpp [hunk 20]
```diff
@@ -0,0 +1,159 @@
+// FluDAG/src/test/test_FlukaFuncs.cpp
```
Comment (ljacobson64) [inline mcnp/test/dagmcnp_unit_tests.cpp]: mcnp not fluka

File: uwuw/tests/uwuw_unit_tests_preprocessor.cpp [hunk 21]
```diff
@@ -215,6 +214,68 @@ TEST_F(UWUWTest,materialMetadata)
...
+
+  virtual void TearDown() {
+    delete uwuw_preproc;
+  }
+};
 
+/*
+ * Test to make sure the name is only uppercased
+ */
+TEST_F(UWUWPreprocTest,testMaterials)
```
Comment (ljacobson64) [inline uwuw/tests/uwuw_unit_tests_preprocessor.cpp:257]: Should there be an "EXPECT_EQ" in either of these tests?

File: common/dagmcmetadata.cpp [hunk 22]
```diff
@@ -0,0 +1,547 @@
...
+        volume_material_property_data_eh[eh] = "mat:Vacuum";
+        volume_material_data_eh[eh] = "Vacuum";
+      } else {
+        volume_material_property_data_eh[eh] = "mat:"+implicit_complement_material;
+        if(implicit_complement_density != "") volume_material_property_data_eh[eh] += "/rho"+implicit_complement_density;
+        volume_material_data_eh[eh] = implicit_complement_material;
+        volume_density_data_eh[eh] = implicit_complement_density;
+      }
+    }
+  }
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp:228]: There should be something in here telling it what to do if no material is found for a volume

**Review by ljacobson64 — CHANGES_REQUESTED** [review]
Made some comments for you to look at.

**Comment by ljacobson64** [comment]
I've tried running a UW2 problem but I've been having a hard time getting uwuw_preproc to work in the first place. When I try the example problem on the website, I get "material Lead was not found in the material library". If I try to run DAGMC anyways, this is what the cell card block on the lcad file looks like:

```
1 Lead  imp:n=1
2 Lead  imp:n=1
3 Lead 12.3 imp:n=1
4 Lead 12.3 imp:n=1
7 0  imp:n=1
8 0  imp:n=1
```

and naturally MCNP segfaults. So I think one of us will need to make sure this works with a UW2 problem before we can merge.

File: common/test/CMakeLists.txt (hunk 11 above)
Comment (makeclean) [reply common/test/CMakeLists.txt:12]: I prefer explicit paths, as we discussed before

File: common/dagmcmetadata.cpp (hunk 8 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: done

File: common/dagmcmetadata.cpp (hunk 7 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: done

File: common/dagmcmetadata.cpp (hunk 6 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp:338]: it should, done :)

File: common/dagmcmetadata.cpp (hunk 4 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: dne

File: common/dagmcmetadata.cpp (hunk 5 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: done

File: common/dagmcmetadata.cpp (hunk 9 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp:256]: It fails, well spotted :)

File: common/dagmcmetadata.cpp (hunk 3 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: done

File: common/dagmcmetadata.cpp (hunk 22 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp:228]: done

File: common/dagmcmetadata.cpp (hunk 2 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: done

File: mcnp/mcnp_funcs.cpp (hunk 1 above)
Comment (makeclean) [reply mcnp/mcnp_funcs.cpp]: done

File: common/dagmcmetadata.cpp (hunk 10 above)
Comment (makeclean) [reply common/dagmcmetadata.cpp]: comment clarified, i think

File: common/test/dagmciface_unit_tests.cpp (hunk 12 above)
Comment (makeclean) [reply common/test/dagmciface_unit_tests.cpp]: Done

File: common/test/dagmciface_unit_tests.cpp (hunk 13 above)
Comment (makeclean) [reply common/test/dagmciface_unit_tests.cpp]: Done

File: common/test/dagmciface_unit_tests.cpp (hunk 14 above)
Comment (makeclean) [reply common/test/dagmciface_unit_tests.cpp]: done

File: fluka/build/CMakeLists.txt (hunk 15 above)
Comment (makeclean) [reply fluka/build/CMakeLists.txt]: Removed

File: geant4/build/CMakeLists.txt (hunk 16 above)
Comment (makeclean) [reply geant4/build/CMakeLists.txt:4]: We still include, its just elsewhere

File: common/CMakeLists.txt (hunk 18 above)
Comment (makeclean) [reply common/CMakeLists.txt:7]: No real reason other than I expect other things to be appended soon,

File: common/CMakeLists.txt (hunk 17 above)
Comment (makeclean) [reply common/CMakeLists.txt]: Nope, removed

File: mcnp/CMakeLists.txt (hunk 19 above)
Comment (makeclean) [reply mcnp/CMakeLists.txt]: Removed

File: mcnp/test/dagmcnp_unit_tests.cpp (hunk 20 above)
Comment (makeclean) [reply mcnp/test/dagmcnp_unit_tests.cpp]: Fixed.

File: uwuw/tests/uwuw_unit_tests_preprocessor.cpp (hunk 21 above)
Comment (makeclean) [reply uwuw/tests/uwuw_unit_tests_preprocessor.cpp:257]: Nope, these are death tests

File: common/test/CMakeLists.txt (hunk 11 above)
Comment (ljacobson64) [reply common/test/CMakeLists.txt:12]: I was mainly saying keep one or the other

Comment (makeclean) [reply common/test/CMakeLists.txt:12]: Oh i see, duh!

**Comment by ljacobson64** [comment]
You said "fixed" on a lot of comments; did you commit your changes?

**Comment by makeclean** [comment]
Not yet, they will be coming, just testing the UWUW stuff now

**Comment by ljacobson64** [comment]
Sounds good

File: common/dagmcmetadata.cpp [hunk 23]
```diff
@@ -0,0 +1,561 @@
...
+      int position;
+      for ( int j = 0 ; j < material_props.size() ; j++ ) {
+        comp_found = material_props[j].find("_comp");
+        if ( comp_found != std::string::npos) {
+          position = j;
+          break;
+        }
+      }
+      // if there is no material property - not failure for impl_comp
+      if(material_props[0] == "" && !(DAG->is_implicit_complement(eh))) {
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: I don't think this block will ever be executed. It's inside the `if( material_props.size() > 1 )` block and I'm pretty sure that means that `material_props[0]` can't be `""`. Perhaps it should be before the `if( material_props.size() > 1 )` block?

Comment (ljacobson64) [reply common/dagmcmetadata.cpp]: I just tested it and if you have a volume with no material, it doesn't give an error and eventually segfaults when MCNP tries to load the lcad file. If you move this `if` statement above this block, then you get the expected behavior; i.e. the `more than one material for volume with id` message and a graceful exit.

**Comment by ljacobson64** [comment]
It will also break if you give it a material # without a density; i.e. `mat:3`. I don't know whether it's worth trying to fix right now or not.

File: common/dagmcmetadata.cpp [hunk 24]
```diff
@@ -183,11 +183,14 @@ void dagmcMetaData::parse_material_data()
...
-      if(try_to_make_int(material_props[0])){
-	std::cout << "Using the simplified nameing scheme without a density" << std::endl;
-	std::cout << "property is forbidden, please rename the group mat:" << material_props[0] << std::endl;
-	exit(EXIT_FAILURE);
-      }
+    }
+
+    // check to see if the simplied naming scheme is used, by try to convert the
+    // material property to an int
+    if(try_to_make_int(material_props[0]) && !(DAG->is_implicit_complement(eh))){
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: I think this is breaking things... if you add a volume to the group `mat:1/rho:-12.3` and then run MCNP, you get

```
Using the simplified nameing scheme without a density
property is forbidden, please rename the group mat:1
```

File: common/dagmcmetadata.cpp [hunk 25]
```diff
@@ -183,11 +183,14 @@ void dagmcMetaData::parse_material_data()
...
+      std::cout << "Using the simplified nameing scheme without a density" << std::endl;
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: naming

File: common/dagmcmetadata.cpp [hunk 26]
```diff
@@ -0,0 +1,602 @@
...
+      grp_name = "mat:"+material_props[0]+"/rho:"+density_props[0];
+      volume_density_data_eh[eh] = density_props[0];
+    } else {
+      grp_name = "mat:"+material_props[0];
+      volume_density_data_eh[eh] = "";
+    }
+
+    // check to see if the simplied naming scheme is used, by try to convert the
+    // material property to an int
+    if(try_to_make_int(material_props[0]) && !(DAG->is_implicit_complement(eh))) {
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp]: I think this is breaking things... if you add a volume to the group `mat:1/rho:-12.3` and then run MCNP, you get

```
Using the simplified nameing scheme without a density
property is forbidden, please rename the group mat:1
```

File: common/dagmcmetadata.cpp [hunk 27]
```diff
@@ -0,0 +1,602 @@
...
+      std::cout << "Using the simplified nameing scheme without a density" << std::endl;
```
Comment (ljacobson64) [inline common/dagmcmetadata.cpp:191]: naming

File: common/dagmcmetadata.cpp (hunk 26 above)
Comment (ljacobson64) [reply common/dagmcmetadata.cpp]: Adding `&& density_props[0].empty()` to this line seems to fix it

**Review by ljacobson64 — APPROVED** [review]

**Comment by ljacobson64** [comment]
Thanks @makeclean !!!!!
